ArcGISRuntimeSDKQt_CppSamples/Layers/ListKmlContents/ListKmlContents.cpp

```diff
@@ -0,0 +1,320 @@
+      m_kmlNodesList << m_kmlDataset->rootNodes()[i];
+
+      buildTree(m_kmlDataset->rootNodes()[i]);
+    }
+
+    // if at top node, then display children
+    if (!m_kmlNodesList.isEmpty() && m_kmlNodesList[0]->parentNode() == nullptr)
+    {
+      displayChildren(m_kmlNodesList[0]);
+    }
+  });
+}
+
+QString ListKmlContents::getKmlNodeType(KmlNode *node)
+{
+  QString type = "";
+  switch (node->kmlNodeType()) {
+  case KmlNodeType::KmlDocument:
+    type = "KmlDocument";
+    break;
+  case KmlNodeType::KmlFolder:
+    type = "KmlFolder";
+    break;
+  case KmlNodeType::KmlNetworkLink:
+    type = "KmlNetworkLink";
+    break;
+  case KmlNodeType::KmlPlacemark:
+    type = "KmlPlacemark";
+    break;
+  case KmlNodeType::KmlPhotoOverlay:
+    type = "KmlPhotoOverlay";
+    break;
+  case KmlNodeType::KmlGroundOverlay:
+    type = "KmlGroundOverlay";
+    break;
+  case KmlNodeType::KmlScreenOverlay:
+    type = "KmlScreenOverlay";
+    break;
+  case KmlNodeType::KmlTour:
+    type = "KmlTour";
+    break;
+  }
+  return " - " + type;
+}
+
+void ListKmlContents::displayChildren(KmlNode *parentNode)
+{
+  if (KmlContainer* container = dynamic_cast<KmlContainer*>(parentNode))
+  {
+    m_levelNodeNames.clear();
+    bool lastLevel = true;
+
+    // for current level, get names of child nodes
+    for (KmlNode* node: *(container->childNodesListModel()))
+    {
+      QString str = node->name() + getKmlNodeType(node);
+
+      // if node has children, add ">" to indicate further levels
+      if (!node->children().isEmpty())
+      {
+        str.append(" >");
+        lastLevel = false;
+      }
+      m_levelNodeNames << str;
+    }
+
+    if (lastLevel)
+    {
+      m_currentNode = container->childNodesListModel()->at(0);
+    }
+    emit levelNodeNamesChanged();
+  }
+}
+
+// recursively build string to indicate node's ancestors
+void ListKmlContents::buildPathLabel(KmlNode* node)
+{
+  if (node->parentNode() != nullptr)
+  {
+    buildPathLabel(node->parentNode());
+    m_labelText.append(">");
+  }
+  m_labelText.append(node->name());
+}
+
+void ListKmlContents::displayPreviousLevel()
+{
+  KmlNode* parentNode = m_currentNode->parentNode();
+  KmlNode* grandparentNode = parentNode->parentNode();
+
+  if (grandparentNode != nullptr)
+  {
+    m_labelText.clear();
+    buildPathLabel(grandparentNode);
+    emit labelTextChanged();
+
+    displayChildren(grandparentNode);
+    m_currentNode = grandparentNode;
+  }
+  // if parent node is nullptr, then at top of tree
+  else
+  {
+    m_labelText.clear();
+    buildPathLabel(parentNode);
+    emit labelTextChanged();
+
+    displayChildren(parentNode);
+    m_isTopLevel = true;
+    emit isTopLevelChanged();
+  }
+
+  if (m_currentNode->name() == "")
+  {
+    m_isTopLevel = true;
+    emit isTopLevelChanged();
+  }
+}
+
+// display selected node on sceneview and show its children
+void ListKmlContents::nodeSelected(QString nodeName)
```

Comment:
looks like `nodeName` should be passed by const&

Reply:
`nodeName` is manipulated with
`nodeName = nodeName.left(ind);` which is why I passed by value

Reply:
but this is doing a copy constructor right? you might as well do:

`QString newNodeName = nodeName.left(ind);`

Reply:
Ohh right! I see